README.md
# serverless-step-functions-local
Run AWS step functions offline with Serverless!

This is a plugin for the [Serverless Framework](https://serverless.com/).  It uses [stepfunctions-localhost](https://www.npmjs.com/package/stepfunctions-localhost) to emulate step functions with AWS' provided tool for local development.

## Requirements

- serverless >= v2.32.0
- The [serverless-offline](https://www.npmjs.com/package/serverless-offline) plugin
- The [serverless-offline-lambda](https://www.npmjs.com/package/serverless-offline-lambda) plugin
- The [serverless-step-functions](https://www.npmjs.com/package/serverless-step-functions) plugin
- Java Runtime Engine (JRE) version 6.x or newer

## Install

`npm install serverless-step-functions-local -D`

## Getting Started

You'll need to add this plugin to your `serverless.yml`.  The plugins section should look something like this when you're done:

```yaml
plugins:
  ...
  - serverless-step-functions
  - serverless-step-functions-local
  - serverless-offline-lambda
  - serverless-offline
  ...
```

Then, add a new section to `config` with `accountId` and `region` parameters:

```yaml
custom:
  stepFunctionsLocal:
    accountId: 101010101010
    region: us-east-1
```

Although not neccessary, it's strongly recomended to add the folder with the downloaded step function executables to `.gitignore`.  By default, this path is `./.step-functions-local`.

The plugin binds to port 8083, this cannot be changed.

It also adds an environment variable for each created state machine that contains the ARN for it.  These variables are prefixed by `OFFLINE_STEP_FUNCTIONS_ARN_`, so the ARN of a state machine named 'WaitMachine', for example could be fetched by reading `OFFLINE_STEP_FUNCTIONS_ARN_WaitMachine`.

## Options

(These go under `custom.stepFunctionsLocal`.)

- `accountId` (required) your AWS account ID
- `region` (required) your AWS region
- `lambdaEndpoint` (defaults to `http://localhost:4000`) the endpoint for the lambda service
- `path` (defaults to `./.step-functions-local`) the path to store the downloaded step function executables
- `TaskResourceMapping` allows for Resource ARNs to be configured differently for local development
- `eventBridgeEvents` allows sending [EventBridge events](https://docs.aws.amazon.com/step-functions/latest/dg/cw-events.html) on execution status changes
  - `enabled` (bool) enabled or disable this feature. Disabled by default.
  - `endpoint` Endpoint for sending events to eg. for [serverless-offline-aws-eventbridge](https://github.com/rubenkaiser/serverless-offline-eventBridge) would be `http://localhost:4010`
- `externalInstance` allows to set a flag to use external instance of local step functions. Disabled by default.
- `stepFunctionsEndpoint` (defaults to `http://localhost:8083`) the endpoint for the step functions service
### Full Config Example

```yaml
service: local-step-function

plugins:
  - serverless-step-functions
  - serverless-step-functions-local
  - serverless-offline-lambda
  - serverless-offline

provider:
  name: aws
  runtime: nodejs10.x


custom:
  stepFunctionsLocal:
    accountId: 101010101010
    region: us-east-1
    TaskResourceMapping:
      FirstState: arn:aws:lambda:us-east-1:101010101010:function:hello
      FinalState: arn:aws:lambda:us-east-1:101010101010:function:hello
    eventBridgeEvents:
      enabled: true
      endpoint: http://localhost:4010
  sqsUrl: http://localhost:4566/101010101010/example-queue

functions:
  hello:
    handler: handler.hello

stepFunctions:
  stateMachines:
    WaitMachine:
      definition:
        Comment: "An example of the Amazon States Language using wait states"
        StartAt: FirstState
        States:
          FirstState:
            Type: Task
            Resource: Fn::GetAtt: [hello, Arn]
            Next: send_message
          send_message:
            Type: Task
            Resource: arn:aws:states:::sqs:sendMessage
            Parameters:
              QueueUrl: ${self:custom.sqsUrl}
              "MessageBody.$": "$"
            Next: wait_using_seconds
          wait_using_seconds:
            Type: Wait
            Seconds: 10
            Next: FinalState
          FinalState:
            Type: Task
            Resource: Fn::GetAtt: [hello, Arn]
            End: true
```
## Running external instance

To run external step functions local using docker, use this command:

```
docker run \
    -p 8083:8083 \
    -e "AWS_ACCOUNT_ID=101010101010" \
    -e "AWS_DEFAULT_REGION=us-east-1" \
    -e "LAMBDA_ENDPOINT=http://localhost:4000" \
    amazon/aws-stepfunctions-local
```

CHANGELOG.md
# 0.1.3

- Allow for resource replacements for local development using custom `TaskResourceMapping` configuration

# 0.2.2

- Allow to use docker instance of step functions local

index.js
const path = require('path');
const StepFunctionsLocal = require('stepfunctions-localhost');
const AWS = require('aws-sdk');
const tcpPortUsed = require('tcp-port-used');
const chalk = require('chalk');
const readLine = require('readline');

class ServerlessStepFunctionsLocal {
  constructor(serverless, options) {
    this.serverless = serverless;
    this.service = serverless.service;
    this.options = options;

    this.log = serverless.cli.log.bind(serverless.cli);
    this.config = (this.service.custom && this.service.custom.stepFunctionsLocal) || {};

    // Check config
    if (!this.config.accountId) {
      throw new Error('Step Functions Local: missing accountId');
    }

    if (!this.config.region) {
      throw new Error('Step Functions Local: missing region');
    }

    if (!this.config.lambdaEndpoint) {
      this.config.lambdaEndpoint = 'http://localhost:4000';
    }

    if (!this.config.path) {
      this.config.path = './.step-functions-local';
    }

    if (!this.config.stepFunctionsEndpoint) {
      this.config.stepFunctionsEndpoint = 'http://localhost:8083';
    }

    if (!this.config.externalInstance) {
      this.config.externalInstance = false;
    }

    this.stepfunctionsServer = new StepFunctionsLocal(this.config);

    this.stepfunctionsAPI = new AWS.StepFunctions({ endpoint: this.config.stepFunctionsEndpoint, region: this.config.region });

    this.eventBridgeEventsEnabled = this.config.eventBridgeEvents && this.config.eventBridgeEvents.enabled;
    if (this.eventBridgeEventsEnabled) {
      this.eventBridgeAPI = new AWS.EventBridge({ endpoint: this.config.eventBridgeEvents.endpoint, region: this.config.region });
    }

    this.hooks = {
      'offline:start:init': async () => {

        if (!this.config.externalInstance) {
          await this.installStepFunctions();
          await this.startStepFunctions();
        }

        await this.getStepFunctionsFromConfig();
        await this.createEndpoints();
      },
      'before:offline:start:end': async () => {
        if (!this.config.externalInstance) {
          await this.stopStepFunctions();
        }
      }
    };
  }

  installStepFunctions() {
    return this.stepfunctionsServer.install();
  }

  async startStepFunctions() {
    let serverStdout = this.stepfunctionsServer.start({
      account: this.config.accountId.toString(),
      lambdaEndpoint: this.config.lambdaEndpoint,
      region: this.config.region
    });

    readLine.createInterface({ input: serverStdout }).on('line', line => {
      console.log(chalk.blue('[Serverless Step Functions Local]'), line.trim());

      if (this.eventBridgeEventsEnabled) {
        this.sendEventBridgeEvent(line.trim());
      }
    });

    // Wait for server to start
    await tcpPortUsed.waitUntilUsed(8083, 200, 10000);
  }

  stopStepFunctions() {
    return this.stepfunctionsServer.stop();
  }

  async getStepFunctionsFromConfig() {
    const parsed = this.serverless.configurationInput;
    this.stateMachines = parsed.stepFunctions.stateMachines;

    if (parsed.custom &&
      parsed.custom.stepFunctionsLocal &&
      parsed.custom.stepFunctionsLocal.TaskResourceMapping
    ) {
      this.replaceTaskResourceMappings(
        parsed.stepFunctions.stateMachines,
        parsed.custom.stepFunctionsLocal.TaskResourceMapping
      );
    }
  }

/**
   * Replaces Resource properties with values mapped in TaskResourceMapping
   */
 replaceTaskResourceMappings(input, replacements, parentKey) {
  for (const key in input) {
    if ({}.hasOwnProperty.call(input, key)) {
      const property = input[key];
      if (['object', 'array'].indexOf(typeof property) > -1) {
        if (input.Resource && replacements[parentKey]) {
          if (typeof input.Resource === 'string' && input.Resource.indexOf('.waitForTaskToken') > -1) {
            input.Parameters.FunctionName = replacements[parentKey];
          } else {
            input.Resource = replacements[parentKey];
          }
        }

        // Recursive replacement of nested states
        this.replaceTaskResourceMappings(property, replacements, key);
      }
    }
  }
}

  async createEndpoints() {
    for (const stateMachineName in this.stateMachines) {
      const endpoint = await this.stepfunctionsAPI.createStateMachine({
        definition: JSON.stringify(this.stateMachines[stateMachineName].definition),
        name: stateMachineName,
        roleArn: `arn:aws:iam::${this.config.accountId}:role/DummyRole`
      }).promise();

      // Set environment variables with references to ARNs
      process.env[`OFFLINE_STEP_FUNCTIONS_ARN_${endpoint.stateMachineArn.split(':')[6]}`] = endpoint.stateMachineArn;
    }
  }

  sendEventBridgeEvent(logLine) {
    let pattern = /(?<date>\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}.\d{3}): (?<executionArn>.+) : (?<data>.+)/;
    let match = pattern.exec(logLine);

    if (match !== null) {
      let eventDate = Date.parse(match.groups.date);

      // Eg. arn:aws:states:us-east-1:101010101010:execution:state-machine-id:execution-id
      let eventExecutionArn = match.groups.executionArn;
      let eventExecutionName = eventExecutionArn.split(':').pop();
      let eventStatemachineArn = eventExecutionArn.replace(':execution:', ':stateMachine:').split(':').slice(0, -1).join(':');

      let eventData = JSON.parse(match.groups.data);

      let eventStatus;
      let eventStartDate = null;
      let eventStopDate = null;

      // https://docs.aws.amazon.com/step-functions/latest/dg/cw-events.html
      // https://docs.aws.amazon.com/step-functions/latest/apireference/API_HistoryEvent.html
      switch (eventData.Type) {
        case 'ExecutionAborted':
          eventStatus = "ABORTED";
          eventStopDate = eventDate;
          break;
        case 'ExecutionFailed':
          eventStatus = "FAILED";
          eventStopDate = eventDate;
          break;
        case 'ExecutionStarted':
          eventStatus = "RUNNING";
          eventStartDate = eventDate;
          break;
        case 'ExecutionSucceeded':
          eventStatus = "SUCCEEDED";
          eventStopDate = eventDate;
          break;
        case 'ExecutionTimedOut':
          eventStatus = "TIMED_OUT";
          eventStopDate = eventDate;
          break;
      }

      if (eventStatus !== undefined) {
        let params = {
          Entries: [
            {
              Detail: JSON.stringify({
                executionArn: eventExecutionArn,
                stateMachineArn: eventStatemachineArn,
                name: eventExecutionName,
                status: eventStatus,
                startDate: eventStartDate,
                stopDate: eventStopDate
              }),
              DetailType: 'Step Functions Execution Status Change',
              Resources: [eventExecutionArn],
              Source: 'aws.states',
              Time: eventDate
            }
          ]
        };

        this.eventBridgeAPI.putEvents(params, function(err, data) {
          if (err) {
            console.error(chalk.bgRed('[Serverless Step Functions Local]'), err, err.stack);
          }
        });
      }
    }
  }
}

module.exports = ServerlessStepFunctionsLocal;
